### PR TITLE
mobile, ethclient, core/types: enable filtering on pending/latest logs

### PR DESCRIPTION
**Pull request by yosriz**
fix a problem getting pending (or latest) logs using `EthereumClient.filterLogs()`, as currently mobile code support only block number.

* wrap rpc.BlockNumber consts for mobile, for use with FilterQuery
* Log - remove 'required' from TxIndex and Index, can be nil for pending
balance

### CODE REVIEW

**Review by yosriz — COMMENTED** [review]
should add negative/positive test to verify the fix (should it be android test?)

File: core/types/gen_log_json.go [hunk 1]
```diff
@@ -70,17 +72,15 @@ func (l *Log) UnmarshalJSON(input []byte) error {
 		return errors.New("missing required field 'transactionHash' for Log")
 	}
 	l.TxHash = *dec.TxHash
-	if dec.TxIndex == nil {
-		return errors.New("missing required field 'transactionIndex' for Log")
```
Comment (YohayBar) [inline core/types/gen_log_json.go]: why not keeping the error if == nil?

File: core/types/gen_log_json.go [hunk 2]
```diff
@@ -70,17 +72,15 @@ func (l *Log) UnmarshalJSON(input []byte) error {
...
+	if dec.TxIndex != nil {
+		l.TxIndex = uint(*dec.TxIndex)
 	}
-	l.TxIndex = uint(*dec.TxIndex)
 	if dec.BlockHash != nil {
 		l.BlockHash = *dec.BlockHash
 	}
-	if dec.Index == nil {
-		return errors.New("missing required field 'logIndex' for Log")
```
Comment (YohayBar) [inline core/types/gen_log_json.go]: same why not keeping the error if == nil?

File: ethclient/ethclient.go [hunk 3]
```diff
@@ -257,6 +257,14 @@ func toBlockNumArg(number *big.Int) string {
...
 		return "latest"
 	}
+	switch number.Int64() {
+	case rpc.PendingBlockNumber.Int64():
+		return "pending"
+	case rpc.LatestBlockNumber.Int64():
+		return "latest"
+	case rpc.EarliestBlockNumber.Int64():
+		return "earliest"
+	}
```
Comment (YohayBar) [inline ethclient/ethclient.go:267]: why is this needed? are you sure it has no side effect in code that use it and expect to get hex string?

Comment (yosriz) [reply ethclient/ethclient.go:267]: this is exactly the fix, json RPC can get either block number or one of these 3 tags. the API exposes just number as block number param, and I used those consts to define those special tags, eventually this code unmarshalled to json, and sent as json RPC.
you can see that the there is similar logic before my added switch, that send 'latest' when number is nil (default param is latest)

Comment (YohayBar) [reply ethclient/ethclient.go:267]: cool 👏

File: core/types/gen_log_json.go [hunk 4]
```diff
@@ -10,16 +10,18 @@ import (
 	"github.com/ethereum/go-ethereum/common/hexutil"
 )
 
+var _ = (*logMarshaling)(nil)
```
Comment (YohayBar) [inline core/types/gen_log_json.go:13]: ? where is it used?

Comment (yosriz) [reply core/types/gen_log_json.go:13]: don't know, it's auto generated file
I needed to autogenerate because of that with pending logs, there are some values that can be null and was required before the fix (tx index and log index)

**Review by YohayBar — APPROVED** [review]
please check "var _ = (*logMarshaling)(nil)" seems it is not used anywhere. otherwise 👍 - please also open a PR to official go-ethereum :)

File: ethclient/ethclient.go [hunk 5]
```diff
@@ -257,6 +257,14 @@ func toBlockNumArg(number *big.Int) string {
 	if number == nil {
 		return "latest"
 	}
+	switch number.Int64() {
+	case rpc.PendingBlockNumber.Int64():
```
Comment (oryband) [inline ethclient/ethclient.go:261]: 1. There is potential loss of information here since `big.Int` might not be `Int64`.
2. See the following code:
```go
// eth/api.go

func (api *PrivateDebugAPI) TraceBlockByNumber(blockNr rpc.BlockNumber, config *vm.LogConfig) BlockTraceResult {
	// Fetch the block that we aim to reprocess
	var block *types.Block
	switch blockNr {
	case rpc.PendingBlockNumber:
		// Pending block is only known by the miner
		block = api.eth.miner.PendingBlock()
	case rpc.LatestBlockNumber:
		block = api.eth.blockchain.CurrentBlock()
	default:
		block = api.eth.blockchain.GetBlockByNumber(uint64(blockNr))
	}
```
```go
// rpc.types.go

type BlockNumber int64

const (
	PendingBlockNumber  = BlockNumber(-2)
	LatestBlockNumber   = BlockNumber(-1)
	EarliestBlockNumber = BlockNumber(0)
)
```
Why not use the original types instead of casting? This feels somewhat hacky...

File: mobile/ethclient.go [hunk 6]
```diff
@@ -23,6 +23,15 @@ import (
...
 	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+const (
+	PendingBlockNumber = int(rpc.PendingBlockNumber)
+
+	LatestBlockNumber = int(rpc.LatestBlockNumber)
+
+	EarliestBlockNumber = int(rpc.EarliestBlockNumber)
```
Comment (oryband) [inline mobile/ethclient.go:34]: These are just declared and not used anywhere in the above code. Why?

**Review by oryband — COMMENTED** [review]
Some questions

File: mobile/ethclient.go (hunk 6 above)
Comment (yosriz) [reply mobile/ethclient.go:34]: these are consts for mobile API, so no use in Go code.
I tried to do the same like other consts defined in mobile API, at this file:
https://github.com/kinfoundation/go-ethereum/blob/master/mobile/accounts.go

File: ethclient/ethclient.go (hunk 5 above)
Comment (yosriz) [reply ethclient/ethclient.go:261]: 1. you're right, number.Int64() might return undefined result if cannot be converted
2. the issue is that the defined consts rpc.types are basically int64, while the input in mobile case is big.Int. so some conversion needed, I can create a big.Int for each case like the following, but that would be wasteful allocations, I might also define the mobile consts as bin.Int for the first place. but they cannot be consts that way.

  ```
  switch number {
	case big.NewInt(rpc.PendingBlockNumber.Int64()):
		return "pending"
	case big.NewInt(rpc.LatestBlockNumber.Int64()):
		return "latest"
	case big.NewInt(rpc.EarliestBlockNumber.Int64()):
		return "earliest"
	}  
```
